Makefile

```diff
@@ -1,16 +1,22 @@
+# THOUGHT: assume this makefile is for local builds and ci would go directly into modules if we want to separate image dependecies
+# eg has Java but no Go? I think this + avoiding unnecessary rebuilds at module levels could allow us to simplify our make structure
+# and speed up builds too
+
 .PHONY: all
 all:
 	$(MAKE) get-deps
 	$(MAKE) build
 
 .PHONY: get-deps
-get-deps:
-	pip install -r requirements.txt
-	$(MAKE) -C master $@
+get-deps: get-deps-pip get-deps-master get-deps-bindings get-deps-webui
 	$(MAKE) -C agent $@
 	$(MAKE) -C proto $@
-	$(MAKE) -C bindings $@
-	$(MAKE) -C webui $@
+.PHONY: get-deps-%
+get-deps-%:
+	$(MAKE) -C $(subst -,/,$*) get-deps
```

Comment:
pretty neat refactor!

**nit** just some newlines between the targets to space them out to be consistent like the other ones.
```
\n
.PHONY X
target:
   blah
\n
.PHONE Y
...
```